src/modules/roc_audio/profiler.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_audio/profiler.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+
+namespace roc {
+namespace audio {
+
+namespace {
+
+const core::nanoseconds_t LogInterval = core::Second;
+
+} // namespace
+
+Profiler::Profiler(core::IAllocator& allocator,
+                   packet::channel_mask_t channels,
+                   size_t sample_rate)
+    : rate_limiter_(LogInterval)
+    , running_data_(allocator)
+    , running_data_head_(0)
+    , running_data_size_(0)
+    , moving_avg_(0)
+    , sample_rate_(sample_rate)
+    , num_channels_(packet::num_channels(channels)) {
+    if (num_channels_ == 0) {
+        roc_panic("profiling writer: n_channels is zero");
+    }
+    if (sample_rate_ == 0) {
+        roc_panic("profiling writer: sample_rate is zero");
+    }
+    running_data_.resize(LogInterval / core::Second * sample_rate_);
+}
+
+void Profiler::begin_frame(Frame& frame) {
+    if (frame.size() % num_channels_ != 0) {
+        roc_panic("profiling writer: unexpected frame size");
+    }
+}
+
+void Profiler::end_frame(Frame& frame, core::nanoseconds_t elapsed) {
+    const double speed = (double)frame.size() / num_channels_ / elapsed * core::Second;
+
+    if (running_data_size_ < running_data_.size()) {
+        running_data_[running_data_head_++] = speed;
+        moving_avg_ = (moving_avg_ * running_data_size_ + speed);
+        moving_avg_ /= (++running_data_size_);
+    } else {
+        moving_avg_ +=
+            (speed - running_data_[running_data_head_]) / (running_data_.size());
+        running_data_[running_data_head_++] = speed;
```

Comment:
It seems there is a problem with this approach: it doesn't take frame size into account.

We're storing last `running_data_.size()` frames, but we don't know what the frame size is and even can't assume that it is the same for every frame.

So, if I understand correctly, when sample rate is 44100 and frame size is 10 ms, we will calculate SMA for last 441 seconds instead of last 1 second.

The difficulty is that the frame size may come from the user of our library when they call `roc_receiver_read()` or `roc_sender_write()` and it may be anything.

Any ideas how to deal with it?

I'm thinking about the following scheme:

* Virtually split LogInterval into fixed-length chunks. Say, 100 chunks of 10 ms each one.
* Track average speed corresponding to each chunk.
* Calculate SMA for these chunks.

So, `running_data_` size is equal to number of chunks (100), and each `running_data_` element is equal to average speed (arithmetic mean) of all *frames* partially or completely *overlapping* with the *chunk* corresponding to the element.

Here is how it looks like:

```
|     chunk 1     |     chunk 2     |     chunk 3     |     chunk 4     |
| frame 1 | frame 2 | frame 3 |           frame 4             | frame 5 |
```

* `chunk 1` will be average of speed for `frame 1` and `frame 2`
* `chunk 2` will be average of speed for `frame 2`, `frame 3`, and `frame 4`
* `chunk 3` will be speed for `frame 4`
* `chunk 4` will be average of speed for `frame 4` and `frame 5`

To implement this, it should be enough to store just a few values:

* current chunk number
* number of samples already added to current chunk (offset in current chunk of the last frame end)
* number of frames already added to current chunk
* average speed for current chunk (arithmetic mean of all frames already added to current chunk)

When a new frame appears, we update the average speed of current chunk and check whether we should switch to the next one.

Reply:
(Note: when computing average speed of a chunk, we likely should weight frames according to the number of samples they contribute (overlap) to the chunk.)

Reply:
So I have been thinking about this algorithm for a while and have realized that after the first second has elapsed and all the chunks are full, the next averages will be gross approximations. We will be removing the average speed of chunk 1 * (elapsed milliseconds) samples from the frame and adding frame.size() new samples to the chunk and recomputing the average. Assuming that there are more than 1 frame in that chunk then the number of samples removed will not be accurate and the next number of samples removed will also not be accurate as the chunks average speed has changed again. Are we okay with this degree of imprecision. 

The alternative mechanism that will be more memory intensive but will be more accurate would be creating a doubly linked list and each node having stored the number of samples and time elapsed. That way we can accurately remove the number of samples that we need to based on that specific frames average and not the approximate average from the chunks.

It would look something like this:
* note each node is a frame stats
* note we are computing average for 25ms to demonstrate the point.

node 1 (samples: 10, time: 10ms) ->  node 2 (samples: 15, time: 5ms) -> node 3 (samples: 20, time: 10ms).

now when we get a new frame that will be node 4 that will for example be node4(samples: 10, time: 5ms). What will happen is that we will chain node4 to the end. Then modify node 1 such that we remove 5 ms worth of samples from it so that node 1 will now look like node1(samples: 5, time: 5ms). So now the linked list still contains 25ms worth of data in it and the samples are accurate then the moving average can be recomputed. Once node1 has time: 0ms then it will be deleted and the tail ptr will point to node2. The moving average in this case will be the sum of all the samples / LogInterval seconds.

Reply:
> So I have been thinking about this algorithm for a while and have realized that after the first second has elapsed and all the chunks are full, the next averages will be gross approximations.

I don't quite understand the problem. Could you provide an example?

How I see it:

```
|   chunk 1   |   chunk 2   |
| frame 1  | frame 2 part 1 | frame 2 part 2 |
```

Assume that our ring buffer is 2-chunks size and the last frame (frame 2) is larger than the remaining "space" (number of samples) in the ring buffer.

In this case frame 2 is split into two parts. The first part is equal to the "space" available in the ring buffer and the second part is the rest. Size of first part is used to weight the frame speed and to add it to chunk 2. Size of the second (remaining) part and speed of the frame are stored in the profiler state.

When profiler will be removing chunk 1 and adding chunk 3, it will find the stored size of the remaining part of last frame (frame 2 part 2) and its speed. It will add the speed weighted with size to the new chunk (chunk 3). If the remaining size is larger than the chunk size, this procedure should be repeated until the remaining part becomes smaller than chunk.

This way each frame will contribute to a chunk proportionally to the size of the part overlapping with the chunk.

Am I missing something?

Reply:
Ah, I understand now. I believe that would work but you would need to keep track of more data than my method. I would suggest we stick to the method I implemented which I find is actually easier to understand and achieves the same outcome.

Reply:
> but you would need to keep track of more data than my method

Well, I wanted to achieve exactly opposite :)

The current code in this PR introduces a linked list of frame nodes. For each read or write operation, we're allocating a node and also probably removing a node. That's a considerable burden to the allocator. We could use a pool (core::Pool) to avoid (re)allocations, but that feels like an overkill for just a frame speed profiler. I'd like to avoid it and use something more lightweight.

On the other hand, I was suggesting the following:

* have a fixed-sized (and rather small) ring buffer, allocated once at session creation
* have a few integer fields for the current profiler state

That's all. No allocations and run-time.

I guess I've failed to to explain the idea clearly, so here is a sample code: https://ideone.com/PEmCwu

It demonstrates how we can fill the ring buffer of fixed-sized chunks from variable-length frames. Then we can compute SMA from that ring buffer, but I omitted that part for brevity.

Reply:
Actually I am still a little confused. I thought we want to profile for the last `LogInterval` seconds. In this case we are profiling the last 500 samples. `CHUNK_LENGTH` is in `nanoseconds_t` but the comment appears like its 50 samples / chunk which won't necessarily elapse in 50 nanoseconds.

Reply:
Ah, sorry, nanoseconds_t is there by mistake. CHUNK_LENGTH should be size_t and it defines number of samples. Operating with samples instead of nanoseconds makes the implementation simpler.

LogInterval should be still defined as nanoseconds because each session can have its own sample rate and thus number of samples will be different. However we can convert from nanoseconds to samples if we know the sample rate and number of channels. These parameters are constant for a session, so we can perform the conversion at construction time.

See also #274, it introduces a helper for such conversion. But until it's merged, we can do it by hand.